lib/ecto/repo/schema.ex

```diff
@@ -112,27 +112,23 @@ defmodule Ecto.Repo.Schema do
 
     ix = case query.select do
       %Ecto.Query.SelectExpr{expr: {:&, _, [ix]}} -> ix
+      %Ecto.Query.SelectExpr{expr: {:merge, _, [{:&, _, [ix]} | _]}} -> ix
       _ -> nil
     end
 
     header = case query.select do
       %Ecto.Query.SelectExpr{expr: {:%{}, _ctx, args}} ->
-        Enum.map(args, fn {field, _} ->
-          case dumper do
-            %{^field => {_, _, false}} -> field
-            %{} -> raise ArgumentError, "cannot select unwritable field `#{field}` for insert_all"
-            nil -> field
-          end
-        end)
+        Enum.map(args, fn {field, _} -> extract_field(field, dumper) end)
+
+      %Ecto.Query.SelectExpr{expr: {:merge, _ctx, merge_args}, take: %{^ix => {_fun, take_fields}}} ->
+        [_, {_, _, merge_fields}] = merge_args
```

Comment:
I believe it can be in the reverse order too like this

```
{:merge, [],
   [
     {:%{}, [], [y: {{:., [type: :binary], [{:&, [], [0]}, :yyy]}, [], []}]},
     {:&, [], [0]}
   ]}
```